Zend/zend_compile.c

```diff
@@ -5381,55 +5447,46 @@ void zend_compile_params(zend_ast *ast, zend_ast *return_type_ast) /* {{{ */
 			arg_info->type = zend_compile_typename(type_ast, default_type == IS_NULL);
 
 			is_class = ZEND_TYPE_IS_CLASS(arg_info->type);
-			arg_type = ZEND_TYPE_CODE(arg_info->type);
+			arg_type = !is_class ? ZEND_TYPE_MASK(arg_info->type) : 0;
 
-			if (arg_type == IS_VOID) {
+			if (arg_type & MAY_BE_VOID) {
 				zend_error_noreturn(E_COMPILE_ERROR, "void cannot be used as a parameter type");
 			}
 
 			if (default_type > IS_NULL && default_type != IS_CONSTANT_AST) {
 				if (is_class) {
 					zend_error_noreturn(E_COMPILE_ERROR, "Default value for parameters "
 						"with a class type can only be NULL");
-				} else switch (arg_type) {
-					case IS_CALLABLE:
+				} else {
+					if (arg_type & MAY_BE_CALLABLE) {
 						zend_error_noreturn(E_COMPILE_ERROR, "Default value for parameters "
 							"with callable type can only be NULL");
-						break;
-
-					case IS_ARRAY:
+					} else if (arg_type & MAY_BE_ARRAY) {
 						if (default_type != IS_ARRAY) {
 							zend_error_noreturn(E_COMPILE_ERROR, "Default value for parameters "
 								"with array type can only be an array or NULL");
 						}
```

Comment:
This case is not necessary. It's handled by "default". Right?

Reply:
This was a different case because the error message is slightly different. I have refactored this code to use a single check and error message format in https://github.com/php/php-src/commit/5dcbc65a53b7f7c6e517b19260c637017528b3f9.